csrc/executor.cpp

```diff
@@ -1889,6 +1889,29 @@ std::vector<at::Tensor> FusionExecutor::evaluateFusionOutputs(
     for (const auto& out_val : fusion()->outputs()) {
       auto out_tensor =
           expr_eval.evaluate(out_val->as<TensorView>()).as<at::Tensor>();
+      if (auto out_tv = dynamic_cast<TensorView*>(out_val)) {
+        const std::vector<IterDomain*>& tv_logical =
+            TensorDomain::noReductions(out_tv->getLogicalDomain());
+        NVF_ERROR(
+            out_tensor.dim() == tv_logical.size(),
+            "ATen-evaluated tensor with shape ",
+            out_tensor.sizes(),
+            " does not match expected dimension of TensorView ",
+            out_tv->toString());
+        for (size_t i : c10::irange(tv_logical.size())) {
+          IterDomain* id = tv_logical[i];
+          if (id->isBroadcast() && !id->hasExpandedExtent()) {
+            NVF_ERROR(
+                out_tensor.size(i) == 1,
+                "Expected size-1 axis at position ",
+                i,
+                " of ATen-evaluated tensor ",
+                out_tv->toString(),
+                " but found size ",
+                out_tensor.size(i));
+          }
```

Comment:
I guess instead of this we could just evaluate the `getMaybeExpandedExtent` for each axis and check that the output size matches. Also, we could wrap this check in `#ifndef NDEBUG`; wdyt @wujingyue ?

Reply:
I like catching errors earlier. Unless it's proven to be expensive, I'd turn it on unconditionally (thus no `#ifndef`).

Question (maybe to @zasdfgbnm): the added checks seem to be a subset of https://github.com/NVIDIA/Fuser/blob/bad998ae277ffc2f43fdc28dca07d01d737a1623/csrc/expr_evaluator.cpp#L99, which is only triggered by `ExpressionEvaluator::bind` at this moment. Is it a bad idea to bind outputs? In the launch-a-kernel path of FusionExecutor::runFusion, `allocateOutputs` binds outputs anyway: https://github.com/NVIDIA/Fuser/blob/bad998ae277ffc2f43fdc28dca07d01d737a1623/csrc/executor.cpp#L1111.

Reply:
> Is it a bad idea to bind outputs?

I don't see any problem for doing so.

Reply:
Thanks guys. Updated to bind the output to `expr_eval`.